### PR TITLE
Add basic test for Chain Growth

### PR DESCRIPTION
**Pull request by nfrisby**
This PR is "part one" for Issue #235.

It adds a `Property` that tests for Chain Growth modulo EBBs and late joins. In particular, it assumes there are no message delays, which is effectively true now and will be until Issue #229 and/or #230.

From https://eprint.iacr.org/2017/573/20171115:001835

> Chain Growth (CG); with parameters τ ∈ (0, 1], s ∈ N. Consider the chains C1, C2 possessed by two honest parties at the onset of two slots sl1, sl2 with sl2 at least s slots ahead of sl1. Then it holds that len(C2) − len(C1) ≥ τ · s. We call τ the speed coefficient.

We test with a deterministic variety of values for the `τ` and `s` parameters. Because our current test framework has a limited set of possible faults, we can test Chain Growth for all values of `s` that are greater than 0. Similarly, we're able to compute a `τ` value for any given interval of slots based on the leader schedule (roughly-- see relevant comments in diff). Once we add message delays (latencies, partitions, etc) these tests as-is will fail -- I'm still thinking about how to handle that, but I've decided to open this PR for now.

### CODE REVIEW

File: ouroboros-consensus/test-consensus/Test/Dynamic/Network.hs [hunk 1]
```diff
@@ -518,23 +559,36 @@ getTestOutput ::
         , NodeConfig (BlockProtocol blk)
         , NodeKernel m NodeId blk
         , NodeInfo blk (StrictTVar m MockFS)
+        , m [(SlotNo, BlockNo)]
+        , m [TraceForgeEvent blk]
+        , m [Point blk]
```
Comment (mrBliss) [inline ouroboros-consensus/test-consensus/Test/Dynamic/Network.hs]: Hmm, I think need a record for these. Maybe even one bundling all seven arguments?

File: ouroboros-consensus/test-consensus/Test/Dynamic/General.hs [hunk 2]
```diff
@@ -170,31 +173,70 @@ prop_general ::
...
-  TestOutput{testOutputNodes} =
-    counterexample ("nodeJoinPlan: " <> condense nodeJoinPlan) $
-    counterexample ("schedule: " <> condense schedule) $
+  TestOutput{testOutputNodes, testOutputSlotBlockNos} =
     counterexample ("nodeChains: " <> unlines ("" : map (\x -> "  " <> condense x) (Map.toList nodeChains))) $
+    counterexample ("nodeJoinPlan: " <> condense nodeJoinPlan) $
     counterexample ("nodeTopology: " <> condense nodeTopology) $
+    counterexample ("slot-node-depth: " <> condense slotNodeDepths) $
+    counterexample ("schedule: " <> condense schedule) $
+    counterexample ("schedule': " <> condense schedule') $
```
Comment (mrBliss) [inline ouroboros-consensus/test-consensus/Test/Dynamic/General.hs]: Hmm, we should be able to come up with a clearer name than `schedule'` (I know it's one of the hardest problems :slightly_smiling_face:).

File: ouroboros-consensus/test-consensus/Test/Dynamic/General.hs [hunk 3]
```diff
@@ -170,31 +173,70 @@ prop_general ::
...
+            b <- Map.lookup s nodeOutputForges
+
+            pure $ not $
+              (||) (nodeIsEBB b) $
+              Set.member (blockPoint b) nodeOutputInvalids
+
+        isFirstJoinSlot s =
+            let NodeJoinPlan m = nodeJoinPlan
+            in
+            Just s == (snd <$> Map.lookupMin m)
```
Comment (mrBliss) [inline ouroboros-consensus/test-consensus/Test/Dynamic/General.hs]: I'd just put this line on the same as `in`

File: ouroboros-consensus/test-consensus/Test/Dynamic/General.hs [hunk 4]
```diff
@@ -220,3 +262,53 @@ prop_general k TestConfig{numSlots, nodeJoinPlan, nodeTopology} schedule
       where
         NumSlots t = numSlots
         NodeJoinPlan m = nodeJoinPlan
+
+    -- check for Chain Growth violations if there are no Common Prefix
+    -- violations
```
Comment (mrBliss) [inline ouroboros-consensus/test-consensus/Test/Dynamic/General.hs]: Can you add a link to the paper + section/page where this is defined? That would be handy.

File: ouroboros-consensus/test-consensus/Test/Dynamic/General.hs [hunk 5]
```diff
@@ -220,3 +262,53 @@ prop_general k TestConfig{numSlots, nodeJoinPlan, nodeTopology} schedule
...
+    --
+    -- We consider all possible non-empty intervals, so the interval span
+    -- @s@ varies but is always at least 1. We compute a different /speed
+    -- coefficient/ @τ@ for each interval under the assumption that there are
+    -- no message delays (ie @Δ = 0@). This is essentially a count of the
+    -- active slots for that interval in the refined @schedule'@.
+    prop_all_growth =
+        (.||.) (not isConsensusExcepected) $
```
Comment (mrBliss) [inline ouroboros-consensus/test-consensus/Test/Dynamic/General.hs]: Using `(.||.)` in prefix notation is a bit cryptic. IMO, infix (albeit with some extra parens in the second operand) would be more readable.

s/..Excepected/..Expected/

File: ouroboros-consensus/test-consensus/Test/Dynamic/General.hs [hunk 6]
```diff
@@ -220,3 +262,53 @@ prop_general k TestConfig{numSlots, nodeJoinPlan, nodeTopology} schedule
...
+        conjoin $ zipWith f slotDepths (List.tails slotDepths)
+      where
+        f ::
+             (SlotNo, BlockNo, BlockNo)
```
Comment (mrBliss) [inline ouroboros-consensus/test-consensus/Test/Dynamic/General.hs]: I think there's a spurious newline here.

File: ouroboros-consensus/test-consensus/Test/Dynamic/General.hs [hunk 7]
```diff
@@ -220,3 +262,53 @@ prop_general k TestConfig{numSlots, nodeJoinPlan, nodeTopology} schedule
...
+          -> [(SlotNo, BlockNo, BlockNo)]
+          -> Property
+        f (s1, _, max1) =
+            conjoin .
+            map (\(s2, min2, _) -> prop_growth (s1, max1) (s2, min2))
+
+        prop_growth :: (SlotNo, BlockNo) -> (SlotNo, BlockNo) -> Property
+        prop_growth (s1, b1) (s2, b2) =
+            counterexample (condense (s1, s2, b1, b2, numActiveSlots)) $
+            (.&&.)
```
Comment (mrBliss) [inline ouroboros-consensus/test-consensus/Test/Dynamic/General.hs]: Same comment about prefix vs infix here.

File: ouroboros-consensus/test-consensus/Test/Dynamic/General.hs [hunk 8]
```diff
@@ -220,3 +262,53 @@ prop_general k TestConfig{numSlots, nodeJoinPlan, nodeTopology} schedule
...
+            [ case map snd bnos' of
+                  [] -> (slot, 0, 0)
+                  o  -> (slot, minimum o, maximum o)
+            | (slot, bnos) <- slotNodeDepths
+            , let bnos' = filter (joinedBefore slot . fst) bnos
+            ]
+
+        joinedBefore slot nid = nodeIdJoinSlot nodeJoinPlan nid < slot
+
+    slotNodeDepths =
```
Comment (mrBliss) [inline ouroboros-consensus/test-consensus/Test/Dynamic/General.hs]: A short explanation of what the slot depth of a node is (and a type signature) would be welcome :slightly_smiling_face:

File: ouroboros-consensus/test-consensus/Test/Dynamic/General.hs [hunk 9]
```diff
@@ -220,3 +262,53 @@ prop_general k TestConfig{numSlots, nodeJoinPlan, nodeTopology} schedule
...
+    -- We consider all possible non-empty intervals, so the interval span
+    -- @s@ varies but is always at least 1. We compute a different /speed
+    -- coefficient/ @τ@ for each interval under the assumption that there are
+    -- no message delays (ie @Δ = 0@). This is essentially a count of the
+    -- active slots for that interval in the refined @schedule'@.
+    prop_all_growth =
+        (.||.) (not isConsensusExcepected) $
+        conjoin $ zipWith f slotDepths (List.tails slotDepths)
+      where
+        f ::
```
Comment (mrBliss) [inline ouroboros-consensus/test-consensus/Test/Dynamic/General.hs]: Can you think of a self-explanatory name for `f`? (I know we use `go` all the time for this :grin:)

File: ouroboros-consensus/test-consensus/Test/Dynamic/Network.hs (hunk 1 above)
Comment (nfrisby) [reply ouroboros-consensus/test-consensus/Test/Dynamic/Network.hs]: I combined these three with `NodeInfo`. Are you OK with the so-called "higher-kinded data" (eg as used in `NodeTracers`)?

File: ouroboros-consensus/test-consensus/Test/Dynamic/General.hs (hunk 2 above)
Comment (nfrisby) [reply ouroboros-consensus/test-consensus/Test/Dynamic/General.hs]: I went with `growthSchedule`. I'm tempted to write `actualSchedule`, but there are some corner cases that make that not quite true. I'll keep my eye out for an opportunity to simplify this as I continue with more PRs in this area.

File: ouroboros-consensus/test-consensus/Test/Dynamic/General.hs (hunk 3 above)
Comment (nfrisby) [reply ouroboros-consensus/test-consensus/Test/Dynamic/General.hs]: Changed to `where`.

File: ouroboros-consensus/test-consensus/Test/Dynamic/General.hs (hunk 4 above)
Comment (nfrisby) [reply ouroboros-consensus/test-consensus/Test/Dynamic/General.hs]: Done; quoted the property too.

File: ouroboros-consensus/test-consensus/Test/Dynamic/General.hs (hunk 5 above)
Comment (nfrisby) [reply ouroboros-consensus/test-consensus/Test/Dynamic/General.hs]: Done

File: ouroboros-consensus/test-consensus/Test/Dynamic/General.hs (hunk 6 above)
Comment (nfrisby) [reply ouroboros-consensus/test-consensus/Test/Dynamic/General.hs]: Still object now that it's named `go`? I'd rather not let the name affect the indent of the type.

File: ouroboros-consensus/test-consensus/Test/Dynamic/General.hs (hunk 7 above)
Comment (nfrisby) [reply ouroboros-consensus/test-consensus/Test/Dynamic/General.hs]: Done; named each conjunct in the `where`

File: ouroboros-consensus/test-consensus/Test/Dynamic/General.hs (hunk 8 above)
Comment (nfrisby) [reply ouroboros-consensus/test-consensus/Test/Dynamic/General.hs]: Done; also renamed to `extrema`.

File: ouroboros-consensus/test-consensus/Test/Dynamic/General.hs (hunk 9 above)
Comment (nfrisby) [reply ouroboros-consensus/test-consensus/Test/Dynamic/General.hs]: Renamed to `go`.

**Comment by nfrisby** [comment]
Thanks for the code critiques. Does the property's semantics make sense?

File: ouroboros-consensus/test-consensus/Test/Dynamic/Network.hs (hunk 1 above)
Comment (mrBliss) [reply ouroboros-consensus/test-consensus/Test/Dynamic/Network.hs]: Yup, HKD is fine for me, the renaming is also good.

File: ouroboros-consensus/test-consensus/Test/Dynamic/Network.hs [hunk 10]
```diff
@@ -478,32 +495,74 @@ directedEdgeInner (node1, LimitedApp app1) (node2, LimitedApp app2) = do
...
+data NodeEvents blk ev = NodeEvents
+  { nodeEventsTipBlockNos :: ev (SlotNo, BlockNo)
+    -- ^ 'ChainDB.getTipBlockNo' for each node at the onset of each slot
+  , nodeEventsForges      :: ev (TraceForgeEvent blk)
+    -- ^ every 'TraceForgeEvent'
+  , nodeEventsInvalids    :: ev (Point blk)
+    -- ^ the point of every 'ChainDB.InvalidBlock' event
+  }
+
+data NodeDBs blk db = NodeDBs
```
[the remark this answers is not in the document]
Comment (nfrisby) [reply ouroboros-consensus/test-consensus/Test/Dynamic/Network.hs]: I just noticed that `blk` is unused in this definition.

[the remark this answers is not in the document]
Comment (nfrisby) [reply ouroboros-consensus/test-consensus/Test/Dynamic/Network.hs]: Done.

File: ouroboros-consensus/test-consensus/Test/Dynamic/Network.hs [hunk 11]
```diff
@@ -478,32 +495,74 @@ directedEdgeInner (node1, LimitedApp app1) (node2, LimitedApp app2) = do
...
+     MonadSTM m
+  => m ( NodeInfo blk (StrictTVar m MockFS) (Tracer m)
+       , m (NodeInfo blk MockFS [])
+       )
+newNodeInfo = do
+  (nodeInfoEvents, readEvents) <- do
+      (t1, m1) <- recordingTracerTVar
+      (t2, m2) <- recordingTracerTVar
+      (t3, m3) <- recordingTracerTVar
+      pure ( NodeEvents     t1     t2     t3
```
[the remark this answers is not in the document]
Comment (nfrisby) [reply ouroboros-consensus/test-consensus/Test/Dynamic/Network.hs]: Done.

File: ouroboros-consensus/test-consensus/Test/Dynamic/General.hs [hunk 12]
```diff
@@ -170,42 +173,81 @@ prop_general ::
...
+    --
+    -- * the node rejected its own new block (eg 'PBftExceededSignThreshold')
+    --
+    -- * the node forged an EBB
+    --
+    growthSchedule :: LeaderSchedule
+    growthSchedule =
+        Map.mapWithKey (filter . actuallyLead) `coerce` schedule
+      where
+        actuallyLead s cid = maybe False id $ do
```
[the remark this answers is not in the document]
Comment (nfrisby) [reply ouroboros-consensus/test-consensus/Test/Dynamic/General.hs]: Done.

**Comment by nfrisby** [comment]
I squashed (totally automatic) all of the commits that arose during review into just one commit.

bors r+

**Comment by nfrisby** [comment]
Minor conflict with the `IOLike` PR -- I'm rebasing now.

**Comment by nfrisby** [comment]
Rebased.

bors retry
